### DISCUSSION TITLE
MOSEK-based CVX versus CVXPY Speed Issue (and CVXPY reduction chain)

### DISCUSSION THREAD

**Discussion by holmesco** [cvxpy/cvxpy#2087] [Q&A] [answered]
I am solving an SDP in CVXPY and have ported to CVX in MATLAB. In both cases I am using MOSEK as the solver.

I noticed that the CVX version solves a lot faster than the CVXPY version. I am fairly sure this has something to do with the reduction chain that is being applied in CVXPY. Most importantly, CVXPY converts to a dualized problem whereas CVX does not.  

Is there a way to force CVXPY to solve the primal or easily bypass the reduction chain (i.e. how do I use the get_problem_data output to solve using MOSEK's python API).

See attached files for MOSEK dump of matlab (CVX) vs python (CVXPY) problems.

[solve_matlab.txt](https://github.com/cvxpy/cvxpy/files/11125658/solve_matlab.txt)
[solve_cvxpy.txt](https://github.com/cvxpy/cvxpy/files/11125659/solve_cvxpy.txt)

**Comment by rileyjmurray** [comment, marked as answer]
CVXPY's conversion to the dual form is unlikely to cause your issue. Our old code didn't dualize but it made the problem _much_ bigger by introducing slack variables.

Have you tried using CVXPY 1.3.1? That release fixed an inefficient step in the canonicalization process, where we were artificially converting sparse matrices to dense matrices.

If you can provide code to generate the CVXPY problem we'd be happy to investigate further.

**Reply by holmesco** [reply]
Thanks for the quick response!

I am using version 1.3.1.  The attached zip file has the python script that runs the CVXPY problem. It also contains the Matlab script so that you can compare (if you have cvx with mosek installed).

[sdp_issue.zip](https://github.com/cvxpy/cvxpy/files/11127812/sdp_issue.zip)

**Reply by rileyjmurray** [reply]
Okay, I don't think the problem is canonicalization inefficiency. Here's the runtime data using MOSEK with my M1 Pro macbook:
```
(CVXPY) Mar 31 08:11:10 PM: Compilation took 4.054e-01 seconds
(CVXPY) Mar 31 08:11:10 PM: Solver (including time spent in interface) took 8.504e+00 seconds
```
So the optimization takes 20x longer than canonicalization.

I also tried using CVXOPT and SCS. Neither of these uses dualization in the way that our MOSEK interface does. Both of them took longer than MOSEK to solve.

The problem is really that CVXPY doesn't canonicalize the problem in the form you've written. It canonicalizes constraints into the "affine" form $\\{ x : A x + b \in K \\}$ where $K$ a convex cone and $A$ is a tall matrix, while the format you've provided is analogous to the "direct" form $\\{ x : A x = b , x \in K \\}$ where $A$ is a wide matrix.

There isn't a nice way around this. In your particular case, you should dualize the high-level CVXPY problem, so that your constraint set is represented in affine form. I tweaked your code to provide the option to do that

```
    if primal:
        """
        min < Q, X >
        s.t.  trace(Ai @ X) == bi, for all i.
        """
        X = cp.Variable(Q.shape, symmetric=True)
        constraints = [X >> 0]
        constraints += [cp.trace(A @ X) == b for A,b in A_b_list]
        cprob = cp.Problem(cp.Minimize(cp.trace(Q @ X)), constraints)
        cprob.solve(solver=sdp_solver,
                    # save_file='solve_cvxpy.ptf',
                    # mosek_params=sdp_opts['mosek_params'],
                    verbose=verbose)
        # Get cost by reversing scaling
        cost_out = cprob.value * adjust[0] + adjust[1]
        X = X.value
    else:
        """
        max < y, b >
        s.t. sum(Ai * yi for all i) << Q
        """
        m = len(A_b_list)
        y = cp.Variable(shape=(m,))
        As, b = zip(*A_b_list)
        b = np.concatenate([np.atleast_1d(bi) for bi in b])
        objective = cp.Maximize(b @ y)
        LHS = cp.sum([Ai * y[i] for (i, Ai) in enumerate(As)])
        constraint = LHS << Q
        prob = cp.Problem(objective, [constraint])
        prob.solve(solver=sdp_solver, verbose=verbose)
        cost_out = prob.value * adjust[0] + adjust[1]
        X = constraint.dual_value
```
All solvers accessed through CVXPY have a much better time trying to solve the dual form. In the case of MOSEK, I got this runtime:
```
(CVXPY) Mar 31 08:33:27 PM: Compilation took 6.034e-02 seconds
(CVXPY) Mar 31 08:33:27 PM: Solver (including time spent in interface) took 7.889e-02 seconds
```

**Reply by holmesco** [reply]
Thanks for the great and in-depth response! I did try to dualize, but had introduced another PSD matrix, which is probably still slowing down the problem. I will change my formulation and report back later.

**Reply by holmesco** [reply]
Formulating the dual as you showed above does make everything work as fast as cvx! Thanks again for the quick and thorough response.

Incidentally, I did start setting up a canonicalization for the "direct" formulation in MOSEK. If I wanted to integrate this into the cvxpy reduction chain, what part of the code would I start looking at?

**Reply by rileyjmurray** [reply]
That's great news! Regarding your question, here's an info dump.

The latest version of MOSEK (v10) supports conic constraints in affine form. I believe one of @phschiele's colleagues was considering updating our MOSEK interface to take advantage of this feature. Once we have this feature it would eliminate the need for our backend dualization. (Note, however, that this would not eliminate the need for _your_ dualization. You need to dualize because all of CVXPY is predicated on the affine formulation, and so no solver can benefit from a direct formulation at the moment.)


But maybe I'm getting ahead of myself. Can you clarify what you mean by the following?
> I did start setting up a canonicalization for the "direct" formulation in MOSEK.

**Reply by holmesco** [reply]
Maybe I am not using the right terminology (canonicalization), but I wrote a script to formulate the "direct" form of the problem in MOSEK. The resulting optimization is just as fast as the dual formulation as well as what I was seeing in MATLAB with CVX. 

I was wondering if it would make sense to add this "direct" formulation into CVXPY, but based on what you said, maybe this conflicts with how the rest of CVXPY parses problems.

**Reply by rileyjmurray** [reply]
You're using terminology fine. I just wanted to clarify. I do think that CVXPY would benefit from supporting more general representations of feasible sets. Unfortunately, making this kind of change would be very involved.

One thing we could do is automatically detect when a conic problem is already in direct form, and then dualize the high-level problem before applying any low-level canonicalization. The dual form would then be amenable to far more efficient representation by CVXPY's "affine" parameterization of feasible sets. @SteveDiamond @phschiele what do you two think?

**Reply by SteveDiamond** [reply]
I had hoped that the new conic mosek interface would handle cases like this. I don't fully understand why the direct form here would inherently be more efficient. Of course solvers handle different formats in different ways, so two mathematically equivalent formulations could lead to quite different solution times. @aszekMosek can you confirm that interfacing to Mosek via the new conic interface will still be slower than the "direct" formulation?

**Reply by rileyjmurray** [reply]
@SteveDiamond even if the new MOSEK interface can handle the affine formulation effectively (I don't think it could), other solvers can't. Right now CVXOPT and SCS struggle with the "primal" form of the problem provided by @holmesco.

**Reply by SteveDiamond** [reply]
Thanks for trying the other solvers! CVX decides based on heuristics whether to dualize the problem before passing it to the solver. Would something like that help?

**Reply by rileyjmurray** [reply]
I think something like that could help a lot. There are some easy enough heuristics for cone programs. It would be fun to work on, but I won't have any bandwidth for that in the near future.

**Reply by Paulnkk** [reply]
@rileyjmurray I would like to tackle that challenge! Are there any papers or resources on the problem? I had considered comparing the ratio of # of variables to # of constraints of dual and primal, but with standard form, the check makes little sense.

**Comment by aszekMosek** [comment]
Hi @holmesco @SteveDiamond @rileyjmurray 

A bit long answer from the point of view of MOSEK.

First of all, it makes a huge difference if you input an SDP in primal or dual form. This is explained in the section https://docs.mosek.com/modeling-cookbook/duality.html#semidefinite-duality-and-lmis especially example 8.7. For this reason if you have something like an SDP relaxation of a QCQP then it should be entered in primal form, whereas a sparse LMI is normally best dualized. The deciding factor is how many constraints you will have that contain semidefinite terms.

Now a fair point would be, that the solver could itself decide whether to dualizeif it pays off (at least heuristically). In fact MOSEK does it for linear and conic problems, but unfortunately not SDP. However, this feature is on the drawing board and should come in not too far future. But for now it is the user's job to input one form or another, and the solver does not attept to dualize it.

CVX and YALMIP are really good at deciding whether to dualize before feeding the problem to MOSEK. So you get a problem that MOSEK solves faster. CVXPY as far as I understand does one fixed thing, so it is good for MOSEK "half of the time", so to speak. It can be seen from the files you attached that CVXPY constructed a problem with a lot more constraints than CVX.

Now all this has *nothing* to do with the interface. The old interface which requires a slack semidefinite variable and the new interface which allows you to say "put this vector of n(n+1)/2 variables in a PSD cone" will ultimately lead to the same thing internally. It is just a change of language which does not affect the main issue which is whether the user entered primal or dual or in other words how big a problem was created. (Which is also why we did not completely understand when CVXPY switched to dualizing everything by default to avoid the slacks which we didn't think were a big issue, but maybe that's a different topic).

**Reply by rileyjmurray** [reply]
Thanks for the response, @aszekMosek! 

To clarify, MOSEK 10, do you internally transform a constraint $A(X) + B \preceq 0$ into $A(X) + B = S$ for an explicit slack variable $S$? If so, then we should probably direct our efforts at smart dualization, rather than rewriting the ``MOSEK(ConicSolver)`` class in CVXPY to use the MOSEK 10 interface.

**Reply by aszekMosek** [reply]
Yes, if you look sufficiently deep down in the optimizer then internally it ends up like this. 

The advantage of the ``Ax+b \in Cone`` interface is simplicity for the user.

**Reply by holmesco** [reply]
@aszekMosek, thanks for the insights regarding problem formulation in MOSEK. They will be very helpful for some of the problems I am working on.

For completeness, I added a version of my problem where I formulate the primal problem directly in MOSEK and compared the runtime of the optimizer only:
- Primal CVXPY:  38.84 s
- Dual CVXPY:     0.12 s
- Primal MOSEK: 0.11 s

For this instance of my problem (which is fairly small) it seems that the primal and dual versions of SDP are not significantly different. Updated code is attached in the zip file.
[sdp_issue.zip](https://github.com/cvxpy/cvxpy/files/11140114/sdp_issue.zip)
